### PR TITLE
Add BaseFailed exceptions for phone_notificator

### PR DESCRIPTION
**Pull request by Konstantinov-Innokentii**
# What this PR does
Introduces BaseFailed exception for phone_notificator.

# Why
We need to somehow distinguish errors we want to be notified - like network errors or invalid twilio credentials (I will call them "real" errors) and errors we want to share with user, but don't want to be paged ( I will call them "fake" errors).
To do that I added "graceful_msg" to all Failed... exceptions.  If details field is present - it mean we can return 400 code with the message, if not - 500 code. So, "real" errors will raise Failed... exception, while "fake" will add "graceful_msg".

# TODO
handle exceptions handled here https://github.com/grafana/oncall/pull/2065
## Checklist

- [ ] Unit, integration, and e2e (if applicable) tests updated
- [x] Documentation added (or `pr:no public docs` PR label added if not required)
- [x] `CHANGELOG.md` updated (or `pr:no changelog` PR label added if not required)

### CODE REVIEW

File: engine/apps/api/views/user.py [hunk 1]
```diff
@@ -650,3 +647,10 @@ def check_availability(self, request, pk):
         user = self.get_object()
         warnings = check_user_availability(user=user, team=request.user.current_team)
         return Response(data={"warnings": warnings}, status=status.HTTP_200_OK)
+
+
+def handle_phone_notificator_failed(exc: BaseFailed) -> Response:
```
Comment (joeyorlando) [inline engine/apps/api/views/user.py:652]:
```suggestion
def handle_phone_notification_failed(exc: BaseFailed) -> Response:
```

File: engine/apps/phone_notifications/exceptions.py [hunk 2]
```diff
@@ -1,24 +1,40 @@
...
+    This exception is indicates error while performing some phone notification operation.
+    Optionally can contain graceful_msg attribute. When graceful_msg is provided it mean that error on provider side is
+    not our fault, but some provider error (number is blocked, fraud guard, ...).
+    By default, graceful_msg is None - it means that error is our fault (network problems, invalid configuration,...).
+
+    Attributes:
+       graceful_msg: string with some details about exception which can be exposed to caller.
+    """
+
+    def __init__(self, graceful_msg=None):
```
Comment (joeyorlando) [inline engine/apps/phone_notifications/exceptions.py:13]: where is `graceful_msg` being set?

Comment (Konstantinov-Innokentii) [reply engine/apps/phone_notifications/exceptions.py:13]: nowhere yet. but it's expected to be set in concrete providers, me or @mderynck will add this

Comment (Konstantinov-Innokentii) [reply engine/apps/phone_notifications/exceptions.py:13]: nowhere yet. but it's expected to be set in concrete providers, me or @mderynck will add this

**Comment by Konstantinov-Innokentii** [comment]
@mderynck could you please add unittests simular to what @joeyorlando did in his [PR](https://github.com/grafana/oncall/pull/2065/files#diff-d8caf3666e2899ad4c0656e5927cf062209cb752fa080dd9fc8a4091507aa5f2R1677)
Otherwise LGTM, thanks for tackling that issue @mderynck and @joeyorlando !

File: engine/apps/twilioapp/phone_provider.py [hunk 3]
```diff
@@ -106,24 +106,51 @@ def finish_verification(self, number: str, code: str):
...
+    20404 - We should not be requesting missing resources
+    30808 - Unknown error, likely on the carrier side
+    30007, 32017 - Blocked or filtered, Intermediary / Carrier Analytics blocked call
+                due to poor reputation score on the telephone number:
+        * We need to register our number or sender with the analytics provider or carrier for that jurisdiction
+    """
+
+    def _get_graceful_msg(self, e, number):
+        if e.code in (30003, 30005):
+            raise f"Destination handset {number} is unreachable"
```
Comment (joeyorlando) [inline engine/apps/twilioapp/phone_provider.py:125]: should this be `return` instead of `raise`?

File: engine/apps/twilioapp/phone_provider.py [hunk 4]
```diff
@@ -106,24 +106,51 @@ def finish_verification(self, number: str, code: str):
...
+            return f"Cannot send to {number} is landline or unreachable carrier"
+        elif e.code == 30410:
+            return f"Provider for {number} is experiencing timeouts"
+        elif e.code == 60200:
+            return f"{number} is incorrectly formatted"
+        elif e.code in (21215, 60410, 60605):
+            return f"Verification to {number} is blocked"
+        elif e.code == 60203:
+            return f"Max verification attempts for {number} reached"
+        return None
```
Comment (joeyorlando) [inline engine/apps/twilioapp/phone_provider.py:138]: nice 👍

File: engine/apps/twilioapp/tests/test_twilio_provider.py [hunk 5]
```diff
@@ -63,3 +70,87 @@ def test_send_sms(mock_messages_create):
...
+
+    with mock.patch(f"apps.twilioapp.phone_provider.TwilioPhoneProvider.{mock_method}") as twilio_mock:
+        twilio_mock.side_effect = TwilioRestException(500, "", code=twilio_code)
+        with pytest.raises(expected_exception) as exc:
+            provider_method(provider)
+            if graceful_msg:
+                assert len(exc.value.graceful_msg) > 0
+            else:
+                assert exc.value.graceful_msg is None
+        twilio_mock.assert_called_once()
```
Comment (joeyorlando) [inline engine/apps/twilioapp/tests/test_twilio_provider.py:151]: 👍

**Review by joeyorlando — APPROVED** [review]

**Pull request merged**